Report a Razor issue feature.

NTaylorMullen opened the pull request:
From a design perspective I'm looking for feedback on how this works. Had to workaround several limitations in VSCode and this is what I came up with. Working on building unit tests for this feature.

- Added the ability to programatically collect usage data in order to make filing Razor issues easier.
- Built a report issue panel that can be interacted with to control how data is collected.
- There wasn't any valid APIs to navigate to GitHub.com's create new issue location with a large payload as the body. Therefore, I took the approach of copying issue content to the clipboard in order to aid in issue creation.
- Expanded the line length limit in tslint. 120 was too restricting and led to hard-to-read code.

How it looks (in two gifs because of upload limitations):
![image](https://i.imgur.com/0DUqU24.gif)
![image](https://i.imgur.com/HQFCwqS.gif)

The full issue content: 
[issueContent.txt](https://github.com/aspnet/Razor.VSCode/files/2691935/issueContent.txt)

#182 

FYI @akshita31 @rchande @danroth27

Eilon commented on src/Microsoft.AspNetCore.Razor.VSCode/package.json:
```diff
@@ -4,13 +4,15 @@
     "vscode-html-languageservice": "^2.1.7",
-    "vscode-languageclient": "^5.1.0"
+    "vscode-languageclient": "^5.1.0",
+    "clipboardy": "1.2.3"
```
OSS good to go.

rchande commented:
@NTaylorMullen Nice, this is a pretty cool idea. We also ran into the GitHub limitation not having a `new issue` API allowing arbitrary length text. I don't remember if we added the ability to automatically copy our generated issues though (@akshita31)? 

It's interesting to me that you need specific gestures to start and stop logging. For us, most issues have a simple repro or are discovered during O# startup (missing SDKs, etc.). 

When this is running in the C# extension, how will we help users understand when they need to use `C#: report an issue` versus `Razor: report an issue`?

akshita31 commented:
For the C# extension we invoke vscode.open on the github new issues url with the body that contains the information from the users system. However it has the restriction that we cannot put the logs etc into it that is why we have to ask users to paste the omnisharp log explicitly.

For the clipboard thing, I remember a discussion to add the capability for the extensions to access the clipboard but i am not sure if it has been added yet : https://github.com/Microsoft/vscode/issues/217

akshita31 commented:
Also the "Start" and "Stop" for the logging collection seems a really cool idea 💯

NTaylorMullen commented:
> It's interesting (but makes perfect sense) to me that you need specific gestures to start and stop logging. For us, most issues have a simple repro or are discovered during O# startup (missing SDKs, etc.).

I wanted to reboot the Razor LanguageServer and start collecting immediately on reboot but I've yet to make a Razor Language Server reboot feasible (it doesn't work properly). If you guys were to have your "start collecting logs" reboot the server and capture those logs it could capture a lot of your concerns. Also, in our case we capture a lot of extra data a user wouldn't typically be able to easily capture such as the C# that VSCode sees vs the C# we see in our data type.

> When this is running in the C# extension, how will we help users understand when they need to use `C#: report an issue` versus `Razor: report an issue`?

Great question, I need to update our Razor log to provide instructions on how to start the file an issue wizard 😄. Today it just says to turn logs to verbose and file an issue on GitHub; can do better now! Also, it might make sense for me to forcibly launch the issue wizard with an error prompt saying "something unexpected happened, could you please repeat what you just did after pressing start?" or something like that when an unexpected error occurs.



> For the clipboard thing, I remember a discussion to add the capability for the extensions to access the clipboard but i am not sure if it has been added yet : [Microsoft/vscode#217](https://github.com/Microsoft/vscode/issues/217)

From my research (prior to using clipbordy) I found that it was added as a proposed API and requires flags from users in order to use it. Therefore, until it's available I decided to use cipboardy since it should be relatively simple to swap out once an official API is available.

ajaybhargavb commented:
Nice idea! I suggest generating collapsible content for pasting in the issue.

<details><summary>CLICK ME</summary>
<p>
```c#
// yay, less noisy issues!
```
</p>
</details>

akshita31 commented:
Also, if you care about mono version(i dont know if razor language server depends on it) you can consider adding that. In omnisharp we have seen some problems specific to a particular mono version.

NTaylorMullen commented:
Note to self: Remember to try this out with light theme. I think my white text will barf

NTaylorMullen commented:
🆙 📅 - No longer in design. Added a ton of tests, changed the test output a bit (attached the new full log output) and verified things don't get unreadable in light mode.

New Issue Content: 
[issueContent.txt](https://github.com/aspnet/Razor.VSCode/files/2701236/issueContent.txt)


Light Mode:
![image](https://user-images.githubusercontent.com/2008729/50320677-52079900-0482-11e9-8fce-b6aae21751f5.png)

NTaylorMullen commented:
Not quite sure whta's happenign with the CI but it says the build can't be found O_o

natemcmaster commented:
/AzurePipelines run all

natemcmaster commented:
Something seems to be busted in the github connection. 🤷‍♂️ I'll try re-authorizing the connection.

natemcmaster commented:
It looks like it was using the old Oauth credentials. I've re-authorized using the GitHub app integration instead.

natemcmaster commented:
/AzurePipelines run all

NTaylorMullen commented:
> It looks like it was using the old Oauth credentials. I've re-authorized using the GitHub app integration instead.

Thanks Nate!

NTaylorMullen commented:
Fixed the build in the PR

SteveSandersonMS commented on client/unittest/ReportIssueCreator.test.ts:
```diff
@@ -0,0 +1,291 @@
+    });
+});
+
+class TestReportIssueCreator extends ReportIssueCreator {
```
I find this pattern a bit odd. We're unit testing methods that are protected only so they can be unit tested. I know we have a habit of doing this in shipping packages where the API surface needs to be locked down, but it's possibly different here since nobody except for us sees these APIs at all.

Other ways we could structure this:

* Only unit test the naturally public API surface. If a behavior isn't observable via public APIs, then there's no reason to specify that behavior.
* Or, maybe make the methods `public`. Nothing apart from our code can even see it either way. Then we don't need the `TestReportIssueCreator` subclass.

I'm not suggesting this is a big problem! If you prefer to leave it as-is, I don't object. I'm just wondering if this is a case where some patterns are leaking out from C#/framework coding practices into a scenario where they aren't really applicable.

NTaylorMullen replied:
> * Only unit test the naturally public API surface. If a behavior isn't observable via public APIs, then there's no reason to specify that behavior.

I try to avoid this approach because you end up writing super complex integration level tests that have to fill out every part of the API that is used. In particular, the VSCode case is extra concerning because a lot of the APIs can't be mocked or faked properly.



> * Or, maybe make the methods `public`. Nothing apart from our code can even see it either way. Then we don't need the `TestReportIssueCreator` subclass.

I made them protected instead of public to avoid production code misuses. Aka, if a new dev is typing dot and then getting a completion list of available methods I don't want them to be using methods that are only public for the sake of testing.

SteveSandersonMS commented on client/unittest/ReportIssueDataCollector.test.ts:
```diff
@@ -0,0 +1,94 @@
+        dataCollector.start();
+
+        // Assert
+        const lastLog = razorOutputChannel[razorOutputChannel.length - 1].trim();
```
Super-minor nitpick: the `.trim()` isn't needed.

Also it's a little odd that the condition on the next line is `> 0` not `>= 0`. Are we deliberately specifying it must not be at the start of the string, or is the intention to check the substring is present anywhere?

I know this is not a big deal.

NTaylorMullen replied:
> Super-minor nitpick: the `.trim()` isn't needed.

👍 

> Also it's a little odd that the condition on the next line is `> 0` not `>= 0`. Are we deliberately specifying it must not be at the start of the string, or is the intention to check the substring is present anywhere?

I'm assuming it's not at the start of the line because we also expect that logs have a date time stamp prefix. That being said, these tests shouldn't be concerned with that so I'll definitely change it for the sake of clarity.

SteveSandersonMS commented on client/unittest/ReportIssueDataCollector.test.ts:
```diff
@@ -0,0 +1,94 @@
+        assert.ok(lastLog.indexOf('Stopping') > 0);
+    });
+
+    it('start->stop->collect captures the last focused Razor document', async () => {
```
It's great to have tests for this kind of thing! Nice one.

SteveSandersonMS commented on src/Microsoft.AspNetCore.Razor.VSCode/src/Diagnostics/ReportIssueCreator.ts:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    // Protected for testing
+    protected sortExtensions(a: vscode.Extension<any>, b: vscode.Extension<any>): number {
```
TBH this feels a bit overcomplicated. Instead of having the `sortExtensions` method and all the unit tests for it, consider replacing L193 with:

```typescript
return extensions.sort((a, b) => 
    a.packageJSON.name.toLowerCase().localeCompare(
        b.packageJSON.name.toLowerCase()
    )
);
```

SteveSandersonMS commented on src/Microsoft.AspNetCore.Razor.VSCode/src/Diagnostics/ReportIssueCreator.ts:
```diff
@@ -0,0 +1,244 @@
+
+        const tableHeader = `|Extension|Author|Version|\n|---|---|---|`;
+        const table = extensions.map(
+            (e) => `|${e.packageJSON.name}|${e.packageJSON.publisher}|${e.packageJSON.version}|`).join(os.EOL);
```
We're using `\n` elsewhere, so it's odd to use `os.EOL` here.

NTaylorMullen replied:
Good catch. A lot of this code was pulled from OmniSharp and I replaced small pieces of it. This was an oversight which was half-replaced.

SteveSandersonMS commented on src/Microsoft.AspNetCore.Razor.VSCode/src/Diagnostics/ReportIssueCreator.ts:
```diff
@@ -0,0 +1,244 @@
+                    if (error) {
+                        throw error;
+                    } else if (stderr && stderr.length > 0) {
+                        throw new Error(stderr);
```
I don't think we should `throw` either here or on L232. Nothing will catch the async exception. I think we should use `reject` instead (also changing L228 to `return new Promise<string>((resolve, reject) => {`).

NTaylorMullen replied:
Interesting, I figured invoking cp.exec in the promise would throw from the inside but I see what you're saying. Good call. I'll fix.

SteveSandersonMS commented on src/Microsoft.AspNetCore.Razor.VSCode/src/Diagnostics/ReportIssuePanel.ts:
```diff
@@ -0,0 +1,170 @@
+                    return;
+                case 'startIssue':
+                    this.issueContent = undefined;
+                    this.dataCollector = this.dataCollectorFactory.create();
```
Should we also check if there was a previous `dataCollector` and if so, dispose it?

Otherwise, if the user clicks "start" repeatedly, they will be leaking `focusRegistration` and `logRegistration`.

As a side note, it's awkward that `start` is separate from the `ReportIssueDataCollector` constructor. It implies you could `start` multiple times, but if you do, that will leak as well. Would it be safer to make `start` implicit in the constructor?

NTaylorMullen replied:
> As a side note, it's awkward that `start` is separate from the `ReportIssueDataCollector` constructor. It implies you could `start` multiple times, but if you do, that will leak as well. Would it be safer to make `start` implicit in the constructor?

hmm, true I could definitely do that. When I wrote the first iteration of this the flow was slightly different which warranted a separate start. Now that's not the case i'll fix.

> Should we also check if there was a previous `dataCollector` and if so, dispose it?
> 
> Otherwise, if the user clicks "start" repeatedly, they will be leaking `focusRegistration` and `logRegistration`.

Great catch, yes we should!

SteveSandersonMS commented on src/Microsoft.AspNetCore.Razor.VSCode/src/Diagnostics/ReportIssueCreator.ts:
```diff
@@ -0,0 +1,244 @@
+                    }
+                });
+            } catch (error) {
+                resolve(`A valid dotnet installation could not be found: ${error}`);
```
Oh actually that's debatable. Maybe you don't want this failure to cause the upstream flow to fail too.

For optimal clarity, I'd recommend using `reject` here, but then also having a `try/catch` around the call to it so the upstream code can explicitly show that it's not going to abort its flow even if `getDotnetInfo` fails.
